### PR TITLE
Batched gradient support for view+inplace operations

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47227 Batched gradient support for view+inplace operations**
* #47365 Add batching rule for torch.clone(tensor, torch.contiguous_format)
* #47364 as_strided batching rule
* #47226 Batching rule for Tensor.new_empty_strided
* #47225 Implement Tensor.new_empty_strided(sizes, strides, *, dtype, device, requires_grad)

Motivation
----------
We would like to compute batched gradients for view+inplace operations.
This most notably shows up in internal implementation of operations.
For example, many view backward functions (SelectBackward, DiagonalBackward)
are implemented with view+inplace, so to support vectorized hessian
computation for e.g. torch.select and torch.diagonal we would need a
way to handle or workaround view+inplace.

Approach
--------
view+inplace creates a CopySlices node and transmute view backward nodes
into an AsStrided node. For example,

```
leaf = torch.randn(4, 5, requires_grad=True)
base = leaf * leaf
view = base[0]
view.cos_()
```

base.grad_fn is CopySlices and view.grad_fn is AsStridedBackward.

To support vmap over CopySlices and AsStridedBackward:
- We use `new_empty_strided` instead of `empty_strided` in CopySlices
so that the batch dims get propagated
- We use `new_zeros` inside AsStridedBackward so that the batch dims get
propagated.

Test Plan
---------
- New tests. When we get closer to having most operations support batched
grad computation via vmap, I'd like to add it as an option to gradcheck
and turn it on for our tests.

Differential Revision: [D24741687](https://our.internmc.facebook.com/intern/diff/D24741687)

### CODE REVIEW

File: test/test_vmap.py [hunk 1]
```diff
@@ -1988,6 +1988,37 @@ def test_slice(self, device):
...
+    @allowVmapFallbackUsage
+    def test_inplace_manyview(self, device):
+        leaf = torch.randn(4, 4, 5, requires_grad=True)
+
+        def func(leaf):
+            # Make sure the function is non-trivially twice differentiable
+            base = leaf * leaf
+            view = base.transpose(0, 2)
+            view = view[1]
+            view = base.diagonal()
```
Comment (albanD) [inline test/test_vmap.py:2014]: Not sure what is the purpose of the first two lines as they get overwritten here.

**Review by albanD — APPROVED** [review]
lgtm!

File: test/test_vmap.py (hunk 1 above)
Comment (zou3519) [reply test/test_vmap.py:2014]: Oops.

**Comment by zou3519** [comment]
> lgtm!

The interesting bits are in the stack below this PR but I'll tag you as reviewer when they are ready (I am just waiting for some tests to pass)

**Comment by albanD** [comment]
Not sure what the black magic in the rest of the stack is. But this one looks surprisingly clean for sure :D

File: test/test_vmap.py (hunk 1 above)
Comment (zou3519) [reply test/test_vmap.py:2014]: Fixed; thanks for the catch!
